openprescribing/matrixstore/db.py

```diff
@@ -0,0 +1,110 @@
+"""
+This module provides the primary interface between the MatrixStore and the rest
+of the application.
+
+It makes use of the assumption that data is static throughout the runtime of
+the application and therefore any data imports or updates will require an
+application restart before they take effect.
+"""
+import os.path
+
+from django.conf import settings
+from django.utils.lru_cache import lru_cache
+
+from frontend.models import Practice
+
+from .grouper import Grouper
+from .open import MatrixStore
+
+
+# Create a memoize decorator (i.e. a decorator which caches the return value
+# for a given set of arguments). Here `maxsize=None` means "don't apply any
+# cache eviction, just keep values for ever"
+memoize = lru_cache(maxsize=None)
+
+
+@memoize
```

Comment:
Why are we accessing the database via a memoized (per-process) file, rather than the Django sqlite backend, which would support concurrency?

Reply:
I've just realised the cache will be shared, so ignore that. 

Though still interested in why we'd not use Django backend.

Reply:
There are few non-standard things we want to do with our SQLite connection including:
 * opening the connection with special flags because we're treating it as read-only which allows SQLite to make various optimisations;
 * setting up our custom SQL functions like `matrix_sum`;
 * deserializing the binary matrix blobs

I'm sure it possible to do all these things within Django by creating a custom database backend and defining custom field types etc, but I suspect it would be a lot more complicated than the code we currently have in `matrixstore.open`.

On top of that, I think the kinds of query we'll want to run are sufficiently non-standard that we wouldn't want to use the ORM for them anyway. And, if anything, having the ORM involved may risk a false familiarity for new team mates which is actively unhelpful.